tests/subprocesstest.py

```diff
@@ -0,0 +1,118 @@
+"""
+subprocesstest enables unittest test cases and suites to be run in separate
+python interpreter instances.
+
+A base class SubprocessTestCase is provided that, when extended, will run test
+cases marked with @run_in_subprocess in a separate python interpreter.
+"""
+import os
+import subprocess
+import sys
+import unittest
+
+
+SUBPROC_TEST_ATTR = '_subproc_test'
+SUBPROC_ENV_VAR = '_SP_TEST'
+
+
+def run_in_subprocess(obj):
+    """
+    Marks a test case that is to be run in its own 'clean' interpreter instance.
+
+    When applied to a TestCase class, each method will be run in a separate
+    interpreter instance.
+
+    Usage on a class::
+
+        from tests.subprocesstest import SubprocessTestCase, run_in_subprocess
+
+        @run_in_subprocess
+        class PatchTests(SubprocessTestCase):
+            # will be run in new interpreter
+            def test_patch_before_import(self):
+                patch()
+                import module
+
+            # will be run in new interpreter as well
+            def test_patch_after_import(self):
+                import module
+                patch()
+
+
+    Usage on a test method::
+
+        class OtherTests(SubprocessTestCase):
+            @run_in_subprocess
+            def test_case(self):
+                pass
+
+
+    :param obj: method or class to run in a separate python interpreter.
+    :return:
+    """
+    setattr(obj, SUBPROC_TEST_ATTR, True)
+    return obj
+
+
+class SubprocessTestCase(unittest.TestCase):
+    def _full_method_name(self):
+        test = getattr(self, self._testMethodName)
+        modpath = test.__module__
+        clsname = self.__class__.__name__
+        testname = test.__name__
+        testcase_name = '{}.{}.{}'.format(modpath, clsname, testname)
+        return testcase_name
+
+    def _run_test_in_subprocess(self, result):
+        full_testcase_name = self._full_method_name()
+
+        sp_test_env = os.environ.copy()
+        sp_test_env[SUBPROC_ENV_VAR] = 'True'
```

Comment:
We should have a comment here saying that `env=` does not inherit from `os.environ`, hence why we are copying and using it.